### ISSUE TITLE
OOM error during json serialisation in exception handling code paths

### ISSUE THREAD

**Issue by stefreak** [garden-io/garden#4592]
### Current Behavior

When Garden tries to serialise cyclic or very large structures, it sometimes fails with "Reached heap limit Allocation failed - JavaScript heap out of memory" error:

<details><summary>Details</summary>
<p>

```
2023-06-09T09:23:10.8797290Z <--- Last few GCs --->
2023-06-09T09:23:10.8797643Z
2023-06-09T09:23:10.8798481Z [2477:0x445f0e0]   896272 ms: Mark-sweep 4047.6 (4138.8) -> 4035.5 (4142.6) MB, 3362.2 / 0.0 ms  (average mu = 0.657, current mu = 0.031) allocation failure; scavenge might not succeed
2023-06-09T09:23:10.8800201Z [2477:0x445f0e0]   900923 ms: Mark-sweep 4051.5 (4142.8) -> 4039.6 (4146.8) MB, 4548.9 / 0.0 ms  (average mu = 0.443, current mu = 0.022) allocation failure; scavenge might not succeed
2023-06-09T09:23:10.8800930Z
2023-06-09T09:23:10.8801435Z
2023-06-09T09:23:10.8801681Z <--- JS stacktrace --->
2023-06-09T09:23:10.8801939Z
2023-06-09T09:23:10.8802351Z FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
2023-06-09T09:23:10.8803062Z  1: 0x9f2b20 node::Abort() [garden]
2023-06-09T09:23:10.8804157Z  2: 0x91b830 node::FatalError(char const*, char const*) [garden]
2023-06-09T09:23:10.8811520Z  3: 0x11c8d6e v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [garden]
2023-06-09T09:23:10.8817784Z  4: 0x11c901a v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [garden]
2023-06-09T09:23:10.8824643Z  5: 0x13b43f5  [garden]
2023-06-09T09:23:10.8833229Z  6: 0x13c3dfd v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [garden]
2023-06-09T09:23:10.8840124Z  7: 0x139dc0d v8::internal::HeapAllocator::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [garden]
2023-06-09T09:23:10.8846845Z  8: 0x139ed94 v8::internal::HeapAllocator::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [garden]
2023-06-09T09:23:10.8853839Z  9: 0x137dc2b v8::internal::Factory::AllocateRaw(int, v8::internal::AllocationType, v8::internal::AllocationAlignment) [garden]
2023-06-09T09:23:10.8861359Z 10: 0x1375bf4 v8::internal::FactoryBase<v8::internal::Factory>::AllocateRawWithImmortalMap(int, v8::internal::AllocationType, v8::internal::Map, v8::internal::AllocationAlignment) [garden]
2023-06-09T09:23:10.8867776Z 11: 0x1377ee6 v8::internal::FactoryBase<v8::internal::Factory>::NewRawOneByteString(int, v8::internal::AllocationType) [garden]
2023-06-09T09:23:10.8874770Z 12: 0x17c4a15 v8::internal::IncrementalStringBuilder::Extend() [garden]
2023-06-09T09:23:10.8882188Z 13: 0x14db255 v8::internal::JsonStringifier::Result v8::internal::JsonStringifier::Serialize_<true>(v8::internal::Handle<v8::internal::Object>, bool, v8::internal::Handle<v8::internal::Object>) [garden]
2023-06-09T09:23:10.8889256Z 14: 0x14dbca2 v8::internal::JsonStringifier::Result v8::internal::JsonStringifier::Serialize_<true>(v8::internal::Handle<v8::internal::Object>, bool, v8::internal::Handle<v8::internal::Object>) [garden]
2023-06-09T09:23:10.8896180Z 15: 0x14dbca2 v8::internal::JsonStringifier::Result v8::internal::JsonStringifier::Serialize_<true>(v8::internal::Handle<v8::internal::Object>, bool, v8::internal::Handle<v8::internal::Object>) [garden]
2023-06-09T09:23:10.8903140Z 16: 0x14dbca2 v8::internal::JsonStringifier::Result v8::internal::JsonStringifier::Serialize_<true>(v8::internal::Handle<v8::internal::Object>, bool, v8::internal::Handle<v8::internal::Object>) [garden]
2023-06-09T09:23:10.8910286Z 17: 0x14dbca2 v8::internal::JsonStringifier::Result v8::internal::JsonStringifier::Serialize_<true>(v8::internal::Handle<v8::internal::Object>, bool, v8::internal::Handle<v8::internal::Object>) [garden]
2023-06-09T09:23:10.8917594Z 18: 0x14dbca2 v8::internal::JsonStringifier::Result v8::internal::JsonStringifier::Serialize_<true>(v8::internal::Handle<v8::internal::Object>, bool, v8::internal::Handle<v8::internal::Object>) [garden]
2023-06-09T09:23:10.8924607Z 19: 0x14dbca2 v8::internal::JsonStringifier::Result v8::internal::JsonStringifier::Serialize_<true>(v8::internal::Handle<v8::internal::Object>, bool, v8::internal::Handle<v8::internal::Object>) [garden]
2023-06-09T09:23:10.8931681Z 20: 0x14ddfd4 v8::internal::JsonStringifier::Result v8::internal::JsonStringifier::Serialize_<false>(v8::internal::Handle<v8::internal::Object>, bool, v8::internal::Handle<v8::internal::Object>) [garden]
2023-06-09T09:23:10.8938627Z 21: 0x14decf8 v8::internal::JsonStringify(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>) [garden]
2023-06-09T09:23:10.8945084Z 22: 0x12589f4 v8::internal::Builtin_JsonStringify(int, unsigned long*, v8::internal::Isolate*) [garden]
2023-06-09T09:23:10.8951174Z 23: 0x1ba0439  [garden]
2023-06-09T09:23:11.8978662Z Command completed after 1 attempt(s).%
```

</p>
</details> 

### Expected behavior

I would expect Garden to fail in a way here that tells us where we can find the problematic code.

### Reproducible example

I cannot reproduce this.

### Workaround

–

### Suggested solution(s)

I would suggest to stop using `JSON.stringify` in error handling code paths, and write our own helper function that uses a javascript generator to stringify the structure and abort if it gets too large, and that has protections in place to avoid OOMing on cyclic structures.

A quick google search revealed for example this library that we could use: https://github.com/Faleij/json-stream-stringify

We should evaluate this (and/or other libraries) to avoid this error in future versions of Garden. 

### Your environment

irrelevant

**Comment by vvagaytsev** [comment]
@mkhq please ping the @garden-io/core when you start working on the stricter types for error details. This issue looks relevant.

**Comment by vvagaytsev** [comment]
We attempted to fix it in #4765. Closing for now, feel free to reopen if it happens again.
